### ISSUE TITLE
What is the replacement for Custom/CustomAsync?

### ISSUE THREAD

**Issue by nvivo** [JeremySkinner/FluentValidation#488]
I have a lot of code like this:

```
CustomAsync(o => {
   // do some logic using multiple fields
   return new ValidationFailure("PropertyThatIsNotInTheModel", "custom error message");
}
``` 

I just updated to the latest version and it tells me to use RuleFor, but I'm not sure how:

```
Rulefor(o => o).MustAsync(o => {
    // return new ValidationFailure won't work here
});
```

So. what exactly is the replacement for the previous code?

**Comment by nvivo** [comment]
Thanks for the quick response.

I wish there could be a simpler way to do that. In some cases, I have lots of logic that return different errors depending on the same flow, and it becomes quite hard to express it with fluent methods. Somethings like this:

````
CustomAsync(o => {
   // do some logic using multiple fields
   if (x)
       return new ValidationFailure("VirtualPropertyA", "custom error message");
   else if (y)  {
      // do more checks on the database
      return new ValidationFailure("VirtualPropertyB", "another error message");
   }
   else {
      // access remote api
      if (...)
          return new ValidationFailure("VirtualPropertyC", "bla bla bla");
   }
}
```

The reason I do that is that some checks may be more expensive, and I don't want to run them if I don't need to. So, OverridePropertyName wouldn't work the same way as it changes depending on the flow.

I'm not sure which were the reasons to remove CustomAsync, which looked quite useful. is there a way to express some other way? 

I'd propose a RuleFor that acts like CustomAsync, but it's not exactly a rule for anything.

**Comment by JeremySkinner** [comment]
As I said, if Custom/CustomAsync work well for you then keep using them....they're marked as deprecated but they're not going to be removed.

**Comment by nvivo** [comment]
Ok, that makes things easier.

But if they're never going to be removed, why are they marked  as obsolete?
It ends up adding lots of warning to the compilation processs, and VS shows them with green markers through the entire file.

**Comment by JeremySkinner** [comment]
Marking something as Obsolete doesn't mean it's going to be removed - it means it's obsolete and not the current/recommended approach. For compatibility, they'll stay in place. 

For most people/use cases it's simpler/cleaner to either use `RuleFor(x=>x)` or custom PropertyValidators to do custom validation. 

If the compiler messages bother you use can always use `#pragma warning disable`

**Comment by nvivo** [comment]
Not trying to argue here, but that's unfortunate. The main reason I use FluentValidation is due to the way it helps me organize this kind of validation, where all other libraries focus too much on field validation and leave the most complicated stuff behind. Having lots of `pragma warning disable` is a bad idea, as it hides a lot of warning for real use cases of Obsolete, that is warning people that some method is going to be removed in the future and there is an alternative.

In this case, there isn't actually an alternative for everything it does.

I hope you may reconsider this decision in the future, or is open to an alternate proposal that suits your guidelines, but still allows to do full model validation without being tied to any specific property.

**Comment by JeremySkinner** [comment]
> Having lots of pragma warning disable is a bad idea, as it hides a lot of warning for real use cases of Obsolete

That's fair enough....but this *is* a real use case for Obsolete. Marking an api as no longer recommended really is the whole point of the attribute. For example there's a lot of obsolete methods in the .net framework but they're not going anywhere - compatibility is really important. FluentValidation is obviously not as big as the .net framework but it still has a lot of users - I'm not going to force a breaking change onto users by changing the public api unless it's absolutely necessary, but methods/apis/approaches evolve over time and the Obsolete attribute is the correct way of notifying users that a certain method is no longer recommended. 

> In this case, there isn't actually an alternative for everything it does.

I really disagree. There are several other ways of doing the same thing. Using `RuleFor(x => x)` with the usual fluent methods is one alternative - you can do anything you would do with Custom, it's just a different approach/mindset. Rather than having the procedural conditionals inside the method, you can use When/Unless combined with OverridePropertyName with the same result.

Another approach would be You could also use a custom PropertyValidator attached using `RuleFor(x => x).SetValidator(new SomeValidator())` as property validators can generate multiple ValidationFailures.

Or you could use `AddRule` and add a DelegateValidator instance (which is what Custom uses internally).
 
So it's up to you - if you want to rewrite it using one of the above approaches then that's most definitely possible, but if you don't to then you'll need to live with the warnings I'm afraid.

**Comment by nvivo** [comment]
I understand your reasoning behind the Obsolete. But I'm not sure if it's the best thing to do here, due to it's side effects. If you're not going to ever remove that method and it serves a valid purpose, a simple comment in the method docs and a recommendation in the wiki would suffice.

I'm not even sure how to rewrite some procedural code with fluent methods. I'd need to pass many temporary variables as part of the model so other methods can make the decision, It becomes a lot of nested method calls that are much harder to read and understand. I tried here and things get ugly very quickly. Using a custom validator would require moving part of the validation logic code to a different class, and that would increase complexity, and since the code is never reused anywhere it makes no sense to have a validator for that. So, in the end, there is no alternative. An alternative would allow me to write custom validation logic in a single block and just change the returns to something different, like adding the error in a context object, or throwing an exception.

Since you're not removing the logic that makes it work for now, what I can do is to inherit AbstractValidator, and copy the code from CustomAsync to my own method and change all my code to use it.

It's just a bummer to have change a lot of valid code for nothing.

**Comment by JeremySkinner** [comment]
> a simple comment in the method docs and a recommendation in the wiki would suffice.

I disagree. Making the method obsolete indicates its use is strongly discourage. A comment doesn't.

> It's just a bummer to have change a lot of valid code for nothing.

Except that you don't actually have to change anything. Or just use AddRule/DelegateValidator directly if the warnings bother you that much.

**Comment by nvivo** [comment]
Ok, so maybe I'm misunderstanding something here. Please advise.

This is what I see in the AbstractValidator for [CustomAsync](https://github.com/JeremySkinner/FluentValidation/blob/master/src/FluentValidation/AbstractValidator.cs#L226):

```
AddRule(new DelegateValidator<T>(x => customValidator(x).Then(f => new[] {f}.AsEnumerable(), runSynchronously: true)));
```

You're saying I should simply replace "CustomAsync" with this everywhere, replacing the `customValidator(x)` with my code. Is this correct?

**Comment by grovesNL** [comment]
I have the same issue as @nvivo. I have many cases that I couldn't find a way to fit cleanly into the other methods. Copy/pasting the `CustomAsync` code into an extension method to hide the warnings seems strange, but I don't want to have warnings littered everywhere.

What is the recommended approach to rewrite async validation rules that don't target a specific property, when the validation needs to be able to build its own validation failure? I use `RuleFor(x => x).MustAsync(method).WithMessage(message)` when possible, but there are many cases where this isn't enough. The approaches described above have been too messy historically, which is why I used `CustomAsync` in the first place.

**Comment by JeremySkinner** [comment]
As mentioned earlier in the thread, if you've got validators using Custom and CustomAsync and this approach works for you then just keep using it.

**Comment by grovesNL** [comment]
To clarify, I'd like to avoid copy/pasting the `Custom`/`CustomAsync` methods into extension methods, or doing this everywhere:
```
#pragma warning disable 618
    CustomAsync(ACustomValidator);
#pragma warning restore 618
```

For now I'll probably create a `CustomAbstractValidator` which inherits `AbstractValidator` to use in its place everywhere, and add the pragma to that class instead. Although this seems like a bad solution just to remove the obsolete warnings.

**Comment by nvivo** [comment]
I'm downgrading to avoid the extra work, and plan to either keep it like that or fork it if I need to fix anything.

Still have no clue why this was marked as obsolete. The old code works fine and causes no problems, The proposed alternative is terrible, there is no way to rewrite all the old code into fluent methods  and the response is basically "don't care".

**Comment by JeremySkinner** [comment]
@nvivo it saddens me that you think I "don't care"...I've taken the time to engage with you and to try and explain the rationale here...I'm sorry if I didn't do a very good job of that, but just because we disagree on something doesn't mean I don't care, and suggesting that really is quite rude considering how much time I spend on this project and answering questions (for free). Instead of complaining, it would be great if you could instead work with me here to come up with something that works for you, while still fitting with the direction I want to take the project. 

So that being said: 

@nvivo @grovesNL, would you have a preference for an alternative api that allows you to return multiple validation failures (like Custom/CustomAsync does), but in a way that fits better with the fluent api? Custom/CustomAsync really doesn't with the style of the rest of the project, so the aim here would be to have something that works similarly to other validators in terms of method chaining, but gives you complete control over the creation of the ValidationFailures. 

An initial thought would be something like this:

```
RuleFor(x=>x).Custom(context => {
  return new[]  {
     new ValidationFailure("SomeProperty", "message");
  };
});
```

This could then exist in a rule chain alongside other validators (and could also be used on non model-level rules). I don't particularly like returning an array here, so an alternative would have some useful helper methods on the context object to clean this up:

```
RuleFor(x=>x).Custom(context => {
   context.AddFailure("SomeProperty", "message");
});
```

Of course you could multiple conditional failures:

```
RuleFor(x=>x).Custom(context => {
   context.AddFailure("SomeProperty", "message");
   if(context.Instance.SomeConditionIsMet) {
     context.AddFailure("SomeProperty", "message");
   }
});
```

I could also envisage AddFailure being clever enough to infer the property name. For example, if using this against a property rather than a model-level rule you could call AddFailure("message") and it would infer the current property unless you choose to override it. 

One downside is I don't think it'd be very clear how something like WithMessage would act on this. e.g. If you call RuleFor.Custom(...).WithMessage("Foo") what should this do? It could just be a no-op, or possibly throw a NotSupportedException. 

Thoughts/suggestions?

**Comment by nvivo** [comment]
@JeremySkinner [I actually asked if you were open to a proposal](https://github.com/JeremySkinner/FluentValidation/issues/488#issuecomment-308467418) that would fit your guidelines, and your answer was clear:
> if you want to rewrite it using one of the above approaches then that's most definitely possible, but if you don't to then you'll need to live with the warnings I'm afraid.

I'm aware of your efforts, and I'm grateful for your work, it's a great project. But I really tried to say your proposal was not helping. Anyway, wenough of that.

The thing with Custom is that it already returns a message and a field that is not typed, so I don't see how you could chain something after that. One of my use cases is that I may return different messages depending on the flow, and that's the reason a `Must` returning bool doesn't work. And you said yourself, having  a `WithMessage` after Custom with a no-op or an exception is just extra maintenance for nothing.

So first, I'd argue that the idea of requiring a fluent method here is flawed. It's trying to impose an ideology while it doesn't solve any problem and it causes others. There is no need to be fluent everywhere, some things just don't map.

I do like the idea of `context.AddFailure("SomeProperty", "message");`. This would make the validation even more powerful to not be limited to a single error. I had use cases for that before, where a single condition would cause similar errors in multiple fields. I vote for replacing `return new ValidationFailure(...)` with `context.AddFailure(...)`.

Now, `RuleFor(x => x)` is just boilerplate for `RuleFor(x => ....` I'm fine with the word `Custom`, but if you want it more similar to RuleFor, maybe `RuleForModel(x => {...})`:

```
RuleForModel((model, context) => {
     context.AddFailure("prop", "message");
});
```

or even:

```
RuleFor((model, context) => {
     context.AddFailure("prop", "message");
});
```

where the overload with 2 parameters would identify this model validation, although I think it would be better to have this method more explicit.

**Comment by JeremySkinner** [comment]
> I actually asked if you were open to a proposal that would fit your guidelines

I apologise - I honestly didn't see that. 

> The thing with Custom is that it already returns a message and a field that is not typed, so I don't see how you could chain something after that.

So the thought here is that you could be defining multiple model-level rules, where some could be defined using the built-in validators and others could not, and require something most custom, eg:

```
RuleFor(x=>x)
  .Must(SomeCheck)
  .WithMessage("something")
  .Custom(context => { ... })
  .SetValidator(new SomeOtherCustomValidator());
```

...and I think this would be a valid scenario. 

I've realised that FV actually already has a solution to the WithMessage no-op. 

The fluent interface has several different components depending on where you are in the chain. For example, when you call RuleFor you receive an `IRuleBuilderInitial`, which allows you to call a validation method (eg NotNull) or SetValidator. You can't call a configuration option (like WithMessage or WithName) at this point -  if you try `RuleFor(x => x.SomeProp).WithMessage("foo")` you won't get intellisense and the code simply wouldn't compile. After you call a validation method or SetValidator, you then receive an `IRuleBuilderOptions`, which allows you to call WithMessage etc on the previously set validator. 

So the solution is actually pretty simple - so long as the Custom method returns IRuleBuilderInitial again instead of IRuleBuilderOptions then you can chain on another validator, but you can't chain on a WithMessage call, which avoids the need for a no-op. 

I'm not opposed to having something like `RuleForModel`, but I think it'd just end up being an alias for `RuleFor(x=>x)` so it's probably not necessary.

**Comment by grovesNL** [comment]
Thanks for reopening this!

The proposal above looks like it would work well for my use case. I agree with the move to `context.AddFailure` because it's probably more logical in complex validation logic anyway. I'm also fine with `RuleFor(x => x)` because I already use that in other areas, so it could be convenient to chain them.

**Comment by nvivo** [comment]
For a compromise, I'm ok with RuleFor(x => x) then. =)

**Comment by JeremySkinner** [comment]
I've pushed 7.1-beta1 to nuget with this change. Please give it a try and let me know if it works OK for you both.  The signature for the Custom  lambda is a `Action<T, CustomContext>`. The signature for CustomAsync's lambda is `Func<T, CustomContext, CancellationToken, Task>`

**Comment by sixeyes** [comment]
I'm trying to understand how to replace my old Custom() call.

Previously I was doing:

`this.Custom(bf => HasIdentifier(bf) ? null : new ValidationFailure(string.Empty, "An identifier must be supplied"));`

I've tried changing this:

`this.RuleFor(bf => bf).Custom(context => HasIdentifier(context) ? null : new ValidationFailure("An identifier must be supplied"));`

But this doesn't compile. Am I close or have I got the wrong end of the stick?

**Comment by JeremySkinner** [comment]
The "new" custom call doesn't return `ValidationFailure` objects directly - you need to add them to the customcontext:

```
RuleFor(bf => bf).Custom((bf, context) => {
  if(HasIdentifier(bf)) context.AddFailure("An identifier must be supplied");
});
```

**Comment by sixeyes** [comment]
Thanks for the quick response. The code compiles but didn't pass my tests. I'd made a typo in my example.
My custom rule is checking that at least one of two properties has been supplied. So I used to call

`new ValidationFailure(string.Empty, "An identifier must be supplied"));`

And in my modified code I forgot to include the `string.Empty` parameter. So I updated the code

`context.AddFailure(string.Empty, "An identifier must be supplied");`

but now I get:
System.ArgumentException : An error message must be specified when calling AddFailure.

I don't understand this.
